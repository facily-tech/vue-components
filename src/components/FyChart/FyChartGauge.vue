<template>
  <div class="gauge-chart d-flex align-end justify-center">
    <svg :height="radius * 1.1" :width="radius * 2">
      <circle
        class="gauge-base"
        :cx="radius"
        :cy="radius"
        fill="transparent"
        :r="innerRadius()"
        :stroke="baseColor"
        :stroke-dasharray="dashArray()"
        stroke-linecap="round"
        :stroke-width="strokeWidth()"
        :transform="transform()"
      />
      <circle
        class="gauge-percent"
        :cx="radius"
        :cy="radius"
        fill="transparent"
        :r="innerRadius()"
        :stroke="percentColor"
        :stroke-dasharray="dashArray()"
        :stroke-dashoffset="offset()"
        stroke-linecap="round"
        :stroke-width="strokeWidth()"
        :transform="transform()"
        style="transition: 'stroke-dashoffset 0.3s'"
      />
    </svg>
    <p class="percent text-h4 font-weight-bold" :style="fontScale()">{{ percent }}%</p>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
  name: 'fy-gauge-chart',

  props: {
    radius: { type: Number, default: 100 },
    percent: { type: Number, default: 0 },
    baseColor: { type: String, default: 'rgb(229, 229, 229)' },
    percentColor: { type: String, default: '#2047f4' },
  },

  methods: {
    strokeWidth(): number {
      return this.radius * 0.2;
    },

    innerRadius(): number {
      return this.radius - this.strokeWidth() / 2;
    },

    circumference(): number {
      return this.innerRadius() * 2 * Math.PI;
    },

    arc(): number {
      return this.circumference() * (180 / 360);
    },

    dashArray(): string {
      return `${this.arc()} ${this.circumference()}`;
    },

    transform(): string {
      return `rotate(180, ${this.radius}, ${this.radius})`;
    },

    percentNormalized(): number {
      return Math.min(Math.max(this.percent, 0), 100);
    },

    offset(): number {
      return this.arc() - (this.percentNormalized() / 100) * this.arc();
    },

    fontScale(): string {
      return `font-size: ${this.radius / 50} + rem !important`;
    },
  },
});
</script>

<style lang="scss">
.gauge-chart {
  position: relative;
  p {
    position: absolute;
    margin: 0 !important;
    line-height: normal !important;
    bottom: 0;
  }
}
</style>
