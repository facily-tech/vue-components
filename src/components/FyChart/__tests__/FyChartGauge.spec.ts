import Vue from 'vue';
import Vuetify from 'vuetify';
import FyChartGauge from '../FyChartGauge.vue';
import { mount, MountOptions, Wrapper } from '@vue/test-utils';

Vue.use(Vuetify);

describe('FyChartGauge.ts', () => {
  // eslint-disable-line max-statements
  type Instance = InstanceType<typeof FyChartGauge>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;
  beforeEach(() => {
    mountFunction = (options?: MountOptions<Instance>) => {
      return mount(FyChartGauge, {
        vuetify: new Vuetify(),
        ...options,
      });
    };
  });

  it('should render component and match snapshot', () => {
    const wrapper = mountFunction();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
