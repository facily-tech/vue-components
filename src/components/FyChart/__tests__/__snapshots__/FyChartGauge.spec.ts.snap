// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FyChartGauge.ts should render component and match snapshot 1`] = `
<div class="gauge-chart d-flex align-end justify-center"><svg height="110.00000000000001" width="200">
    <circle cx="100" cy="100" fill="transparent" r="90" stroke="rgb(229, 229, 229)" stroke-dasharray="282.7433388230814 565.4866776461628" stroke-linecap="round" stroke-width="20" transform="rotate(180, 100, 100)" class="gauge-base"></circle>
    <circle cx="100" cy="100" fill="transparent" r="90" stroke="#2047f4" stroke-dasharray="282.7433388230814 565.4866776461628" stroke-dashoffset="282.7433388230814" stroke-linecap="round" stroke-width="20" transform="rotate(180, 100, 100)" class="gauge-percent" style="transition: 'stroke-dashoffset 0.3s';"></circle>
  </svg>
  <p class="percent text-h4 font-weight-bold">0%</p>
</div>
`;
